### PR TITLE
Extract Ruby extension

### PR DESCRIPTION
**Pull request by vitallium**
This PR extracts Ruby and ERB support into an extension and removes the built-in Ruby and Ruby support from Zed.

As part of this, the new extension is prepared for adding support for the `Ruby LSP` which has some blockers. See https://github.com/zed-industries/zed/pull/8613 I was thinking of adding an initial support for Ruby LSP but I think it would be better to start with extracting the Ruby extension for now.

The implementation, as the 1st step, matches the bundled version but with 3 differences:

1. Added signature output to the completion popup. See my comment below.
CleanShot 2024-05-04 at 09 17 37@2x

3. Use the shell environment for starting the `solargraph` executable.  See my comment below.
4. Bumped the tree sitter version for Ruby to the latest available version.

Additionally, I plan to tweak this extension a bit in the future but I think we should do this bit by bit. Thanks!

Release Notes:

- Removed built-in support for Ruby, in favor of making it available as an extension.

### CODE REVIEW

File: extensions/ruby/extension.toml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+authors = ["Vitaly Slobodin <vitaliy.slobodin@gmail.com>"]
+repository = "https://github.com/zed-industries/zed"
+
+[language_servers.solargraph]
+name = "Solargraph"
+language = "Ruby"
+
+[grammars.ruby]
+repository = "https://github.com/tree-sitter/tree-sitter-ruby"
+commit = "9d86f3761bb30e8dcc81e754b81d3ce91848477e"
```
Comment (vitallium) [inline extensions/ruby/extension.toml:15]: Bumped the grammar version from `0.20.0` to the latest version  
https://github.com/tree-sitter/tree-sitter-ruby/commit/9d86f3761bb30e8dcc81e754b81d3ce91848477e

File: extensions/ruby/src/language_servers/solargraph.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                    _ => return None,
+                };
+
+                let len = completion.label.len();
+                let name_span =
+                    CodeLabelSpan::literal(completion.label, Some(highlight_name.to_string()));
+
+                Some(CodeLabel {
+                    code: Default::default(),
+                    spans: if let Some(detail) = completion.detail {
```
Comment (vitallium) [inline extensions/ruby/src/language_servers/solargraph.rs:52]: Added signature output to the completion:

CleanShot 2024-05-04 at 09 17 37@2x

File: extensions/ruby/src/ruby.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        worktree: &zed::Worktree,
+    ) -> Result<zed::Command> {
+        match language_server_id.as_ref() {
+            Solargraph::LANGUAGE_SERVER_ID => {
+                let solargraph = self.solargraph.get_or_insert_with(|| Solargraph::new());
+
+                Ok(zed::Command {
+                    command: solargraph.server_script_path(worktree)?,
+                    args: vec!["stdio".into()],
+                    env: worktree.shell_env(),
```
Comment (vitallium) [inline extensions/ruby/src/ruby.rs:30]: To pick up the correct Ruby version for a project, version managers like `asdf`, `mise` or `rbenv` require to use shell environment.

File: extensions/ruby/src/language_servers/solargraph.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            }
+            _ => None,
+        }
+    }
+
+    pub fn label_for_symbol(&self, symbol: Symbol) -> Option<CodeLabel> {
+        let name = &symbol.name;
+
+        return match symbol.kind {
+            SymbolKind::Method => {
```
Comment (vitallium) [inline extensions/ruby/src/language_servers/solargraph.rs:72]: The implementation here is what I have been struggling a lot with. I ended up with using `literals` to highlight types properly, I hope this is ok.

File: extensions/ruby/src/language_servers/solargraph.rs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                let display_range = class.len()..class.len() + name.len();
+
+                Some(CodeLabel {
+                    code,
+                    spans: vec![CodeLabelSpan::code_range(display_range)],
+                    filter_range: filter_range.into(),
+                })
+            }
+            SymbolKind::Constant => {
+                let code = format!("{}", name.to_uppercase());
```
Comment (vitallium) [inline extensions/ruby/src/language_servers/solargraph.rs:108]: In Ruby, `constants` start with an uppercase letter, but I didn't want to make the implementation here more complicated, so uppercasing the `constant` should be good too

File: extensions/ruby/LICENSE-APACHE
Comment (maxdeviant) [inline extensions/ruby/LICENSE-APACHE:1]: Can we change this file to be a symlink to `LICENSE-APACHE` in the repository root?

Comment (vitallium) [reply extensions/ruby/LICENSE-APACHE:1]: Sure. Changed it. Thanks for noticing that!

**Comment by maxdeviant** [comment]
@vitallium Thanks for putting this together!

This looks like a very good start to a Ruby extension. We can continue to iterate on the completions/symbol display, as needed.

**Comment by vitallium** [comment]
> @vitallium Thanks for putting this together!
> 
> This looks like a very good start to a Ruby extension. We can continue to iterate on the completions/symbol display, as needed.

Thanks for helping me here and reviewing this PR!

**Pull request merged**